spec/compromise-syntax.md

```diff
@@ -0,0 +1,238 @@
+For example, certain strings may need to be inlined as literals so that
+they are not changed during translation.
+Numeric constants need to be formatted differently depending on the target language
+(e.g., which digits and separators, and the grouping style).
+Date constants need to be formatted according to the target language’s calendar system.
+
+If only a value literal is given, without specifying a function,
+then its string value is used verbatim and it is read-only for translators.
+- TODO: Value literals need to be delimited (they may contain spaces),
+  and the starting delimiter needs to be distinct from the prefixes for
+  argument names and functions.
+  Reasonable choices include `<>`, `()`, `[]`, or `||`.
+  Consider that the same delimiters should also be usable (not visually confusing)
+  when used in a list of selection values (see below); that probably excludes `||` and `[]`.
+- TODO: Define escaping inside constant values.
+  Probably the pattern escapes plus escapes for the constant delimiters.
+
+A placeholder must not be an empty pair of `{}` braces.
+
+Any character that does not fit defined syntax is an error.
+This leaves room for future extensions.
+For example, a placeholder must start with `{` immediately followed by
+the prefix character for an argument name, literal value, or function;
+and after the function name there must be only white-space-separated options which
+start with identifier-start characters.
+
+# Syntactic white space
+
+We use white space inside placeholders and in “code mode” (outside patterns) as token separators.
+White space is a sequence of one or more of the characters TAB, LF, CR, SP, and maybe some more.
+- TODO: For the definition of white space we should consult with the Unicode Source Code Working Group.
+- TODO: Decide whether to use Unicode Pattern_White_Space or otherwise allow RLM and LRM characters.
+
+White space can also be useful for line breaking long messages, indentation, and alignment.
+However, we should not allow white space everywhere possible,
+because that just leads to confusing variations in style,
+and the creation of formatting tools to enforce certain styles.
+For example, there is no reason to allow white space between a name or function prefix and its identifier,
+around the `=` of an option, after the `{` of a placeholder, or before the `}` of a placeholder.
+
+# Pattern selection
+
+Messages need the ability to choose among variants of a pattern based on certain argument values.
+Common examples include selecting the right plural form, and variants for different person genders.
+
+There should be a single level of selection (not nested like in ICU MessageFormat).
+It needs to support multiple selectors.
+
+In this syntax, a list of N selectors is followed by a list of pairs where
+the first element of each pair is a list of N value literals and
+the second element of each pair is a pattern.
+A `_` is a wildcard value that always matches.
```

Comment:
I'd prefer `*` to `_`.

I would tend to say:

> A `_` is a wildcard value that matches any otherwise unmatched value.

Reply:
I don't care. This is another item that I have seen used by the committee. `*` works for me. ICU MessageFormat has been using `other` for its select/plural/selectordinal.

Reply:
Changed to `*` as suggested in https://github.com/unicode-org/message-format-wg/issues/270